### ISSUE TITLE
Several issues with Specifications v1.3

### ISSUE THREAD

**Issue by avih** [tldr-pages/tldr#4290]
Some non exhaustive issues with spec v1.3:

Terminology:

- You link to the terminology RFC, but use terms which are not specified there, like "MAY NOT". It's not obvious what it means (one could guess, but the capitalization suggests it should be specified and guessing should not be required). If you specify terminology, stick to it and to its explicit semantics.

Client program:

- The overall syntax of the client is not mentioned anywhere. I'm guessing it's possibly `client [OPTIONS] CMD` - which suggests that CMD is mandatory, but judging by other text at the page, that's not necessarily the case.

- While it's specified what some options should do, it's not mentioned anywhere what the client should do when no options are used (but CMD is). I'm assuming the main use case it to print or display some form of the page of CMD, but it's not mentioned anywhere.

- It's not mentioned how options may or may not be combined. For instance `-u` with `CMD` could be useful, but `-l` with `CMD` is probably not useful, while `-v` probably doesn't go with anything. Should it succeed without arguments at all?

Terminal output:

- You give a (good) example how commands should be listed, but the rest is vague. Specifically, when it comes to the main use case (which I assume is) `client CMD` it doesn't say if only colors should be stripped, or maybe also the formatting of markup-to-display - i.e. print a raw page file. This goes back to the main task of the client - which isn't specified.

Languages/locales:

I'm ignoring issues with the current directory structure, and assume #4120 is the way forward, at which case it's not obvious to me what the algorithm should be to map a locale string to directory name string, but I think it should be specified in absolute terms in order to be useful, because the example at #4120 doesn't show directories with coutry codes.

One simplistic and most probably bad algorithm could be:
- Strip the country and anything else beyond the language code, and search `pages/<language>`.
- If language is not `en`, then also search `pages/en`.

E.g. both `en_US.UTF-8` and `en_UK` map to `pages/en`, `zh_CN` and `zh_TW` and `zh` all map to `pages/zh`, etc.

Another algorithm could be to seatch `pages/lc_CC` and then `pages/lc` and then (if `lc` is not `en`) also `pages/en`.

Now, I don't pretend to know much about locales and my ideas is probably complete junk. But you should define absolutely how a locale string maps to directory name string(s) which should be searched, because otherwise it's unlikely IMHO that clients will be able to get the directory name as you expect them to.

More specific locale issues:
- Should `LC_ALL` override other vars? (generally it does, but you don't mention it).
- This doesn't sound useful to me other than for development purposes: "If such a command-line option is specified, a client must strictly adhere to its value, and MUST NOT show pages in a different language", because it means it will not fall back to English if the page was not found in the specified language, which I don't think is what the user was hoping for. It also somewhat contradicts the earlier directive of "clients MUST always attempt to fallback to English if the page does not exist in the user preferred language". If the command line option overrides the fallback, then I think it should be mentioned. However, as far as expectations go, I think the user _would_ like it to fall back to english.

**Comment by sbrl** [comment]
Thanks for the feedback, @avih! I've responded to each of your invidivdual points below.

> You link to the terminology RFC, but use terms which are not specified there, like "MAY NOT". It's not obvious what it means (one could guess, but the capitalization suggests it should be specified and guessing should not be required). If you specify terminology, stick to it and to its explicit semantics.

Oh, great point - thanks! We should replace those - e.g. `MAY NOT` with `MUST NOT`.

----------------------

> The overall syntax of the client is not mentioned anywhere. I'm guessing it's possibly client [OPTIONS] CMD - which suggests that CMD is mandatory, but judging by other text at the page, that's not necessarily the case.

True. We could perhaps use a section in the spec for that. It needs to be worded such that it's only applicable to CLI clients though.

--------------------

> While it's specified what some options should do, it's not mentioned anywhere what the client should do when no options are used (but CMD is)

Another interesting edge case. I would guess that some help text should be shown. We should update the spec.

------------------

> It's not mentioned how options may or may not be combined. For instance -u with CMD could be useful, but -l with CMD is probably not useful, while -v probably doesn't go with anything. Should it succeed without arguments at all?

That could get complicated. We should perhaps mention which options ar  mutually exclusive with which other ones, but the trick there is to find a concise way of expressing that without making the spec look too verbose.

---------------------

> You give a (good) example how commands should be listed, but the rest is vague. Specifically, when it comes to the main use case (which I assume is) `client CMD` it doesn't say if only colors should be stripped, or maybe also the formatting of markup-to-display - i.e. print a raw page file. This goes back to the main task of the client - which isn't specified.

This is very client-dependent. Some clients will support colour, and others don't. The key thread that links them all is the support of CommonMark though, as the spec says:

> Although this specification is about the interface that clients must provide, it is also worth noting that pages are written in standard CommonMark, which the exception of the non-standard {{ and }} syntax, which surrounds values in an example that users may edit. Clients MUST NOT break if the page format is changed within the CommonMark specification.

---------------------------------

> at which case it's not obvious to me what the algorithm should be to map a locale string to directory name string

The spec is actually very clear on this. POSIX style locale codes are used. The spec details that [here](https://github.com/tldr-pages/tldr/blob/master/CLIENT-SPECIFICATION.md#translations):

> The format of these directories is `pages.<locale>`, where `<locale>` is a [POSIX Locale Name](https://www.gnu.org/software/gettext/manual/html_node/Locale-Names.html#Locale-Names) in the form of `<language>_<country>`, where:

------------------------

Hrm, that approach has a number of issues. Regional dialects can have significant variances, and the resolution order for different languages is already clearly defined in the [languages section](https://github.com/tldr-pages/tldr/blob/master/CLIENT-SPECIFICATION.md#language) of the spec. Specifically, it's the `LANG` and `LANGUAGE` environment variables that are used.

-------------------------

> Should LC_ALL override other vars? (generally it does, but you don't mention it).

Oh, oops! We've actually [discussed that before](https://github.com/tldr-pages/tldr/issues/4092#issuecomment-647021447), but it must have been dropped. Yeah, `LC_ALL` shoudl override `LANG` and `LANGUAGE` - this will ensure that we're fully POSIX compliant.

--------------------

> This doesn't sound useful to me other than for development purposes

I think that's the intention. Then for example we can request that a user checks with the `-L` argument that a page is present in that language. We actually have this problem a lot with platforms (some clients don't follow the full page resolution algorithm), so users open issues to request pages when they already exist. I anticipate that this will become an issue with languages too, so a quick way to override it and specify a specific language would be very useful.

--------------------

Many thanks again for all the comments! Would oyu like to open a PR to address some or all of these comments?

**Comment by avih** [comment]
Many thanks for the detailed reply!

> Oh, great point - thanks! We should replace those - e.g. `MAY NOT` with `MUST NOT`.

I would not do that blindly. As far as I can tell it's actually used as "not required" - exactly demonstrating why correct terminology is important. Excellent example: "Clients MAY NOT support new platforms (though such support is RECOMMENDED)", which could be changed, for instance, to "It's RECOMMENDED that clients support new platforms automatically (without changes to the client)". That's it, RECOMMENDED already includes not-required by the RFC specification, and it's also stronger than MAY.



> It needs to be worded such that it's only applicable to CLI clients though.

I think the spec already does that quite a lot, so I'm not seeing it as a blocker.



>> it's not mentioned anywhere what the client should do when no options are used (but *CMD is*)

> Another interesting edge case. I would guess that some help text should be shown

It's not an edge case, it's the main case. `client CMD`. No options, just CMD.


> That could get complicated. We should perhaps mention which options ar mutually exclusive...

POSIX has a great set of recomendations how utilities should specify their syntax in a non-ambiguous way - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html



> The spec is actually very clear on this. POSIX style locale codes are used. The spec details that [here](https://github.com/tldr-pages/tldr/blob/master/CLIENT-SPECIFICATION.md#translations):

That's the spec for the current directory structure, which I didn't want to refer to because my impression was that it's being moved to another structure - as mentioned at #4120.

But if only the structure changes while the names remain in the same format as now, then for instance the locale `it_IT` would not find any pages, because the only Italian pages are in `pages.it` and not `pages.it_IT`.

> Hrm, that approach has a number of issues. Regional dialects can have significant variances

I specifically mentioned I don't know much about locales, and it's probably a very bad algorithm. It was given only as an example of what I mean by explicit algorithm.

So together with the spec mentioned earlier, It definitely makes it easier for clients to *search* pages, but I think practically also a lot easier to fail to *find* pages which the locale does suggest should be found - like the example with `it_IT` and `it`. Do people in italy have `it` (without the `_IT` part) in their LANG or LANGUAGE? (I actually don't have a clue, but most search results from coursory search do suggest that `it_IT` is the common form).

> so a quick way to override it and specify a specific language would be very useful

Fair enough. In that case I think It would help to mention just that. Otherwise it can *appear* contradicting that in one place it states that the client MUST fall back to english, and few lines below that it MUST strictly afhere to the specific language. So at "clients MUST always attempt to fallback to English..."  maybe add "except with -L is used - see below" or some such.

> Would oyu like to open a PR to address some or all of these comments?

Not right now. Getting good specifications right actually requires considerable effort. I mostly pointed at issues, which is a good start, but could still be far from finding the solution, because the more accurate you try to be, the more you notice issues, which also require resolution and more specifications. It's not a trivial process.

I actually noticed these issues because I started to implement the specifications, and had so many questions which were not answered by it. So I wanted to check how other clients behave, but from the few clients I tried (node.js, python, bash, sh, perl) it doesn't look like any of them is actualluy compliant... so I was left with the spec alone, which left many things unanswered for me...

Thanks again for the reply. Feel free to ping me if you think I could help with some things (and obviously I'll keep watching this thread).

**Comment by avih** [comment]
@sbrl I'm still not entirely clear on the language-to-directory resolution clients are expected to use. Do you prefer that I open a new issue for it?

**Comment by avih** [comment]
Also on the subject of missing basic client behavior description.

Say someone who tries to implement the spec makes an educted guess that `<client> NAME` should display the help page for command NAME. They skim the spec, then reach this line:

> Clients must not assume that a given command is always executable on the host platform.

and they go "wait, WHAT?!?", was I supposed to run the command NAME? why else would they tell me not to assume it's executable?! was I even supposed to check if it exists at all? They read the spec again and realize no where does it say what the client should actually do.

The spec could have said with equal validity: "clients must not make coffee without cream". This makes no sense at all, unless it was specifed earlier, for instance, that `<client> NAME` should produce a cup of coffee with a sticker NAME on it.

So don't assume it's obvious what a client should do, and then go on to specify in gread details what some options must or must not do, and then tell it what NOT to do when you never tell it what YES to do.

And another subject, it's technically impossible to "detect" the current platform without knowing IN ADVANCE the exact list of platforms which you have pages for - and even that requires considerable guessing of matching the current platform to one or more of the platforms which you support.

For instance, the strings "osx" or "windows" etc are not produced by any API which a client could use. In fact, the only platform which you have (currently) and can be identified automatically by a client without knowing the tldr platforms in advance is `linux` (`uname`).

Even if the client is a macOS program which can only be run on macOS, it still can't know that it should look for it in `pages/osx/...` without knowing in advance that you have `osx` in your "platforms". If you don't specify that you have `osx` in your platforms, then you might as well called the platforms `A`, `B`, `C` and `D`. A client just can't know what YOUR choice for dir name was if you don't specify it - and which general group of platforms it covers. Don't assume it's obvious because it's very very far from obvious.

**Comment by avih** [comment]
Here are my suggestion to improve some of the issues mentioned above, not orgnized in specification form. I think that's it for now.

- These specifications target non-interactive client programs which can be invoked from the command line.

- A client may offer more options or support other use cases than these specifications require, as long as they don't conflict with these specifications.

- A client which mostly complies should document the conditions and exceptions for its compliance. For instance an interactive client could say that it supports the options specified below, but it bcomes interactive (and outside this specifications) instead of exiting. Or that it complies fully when its standard input/output is not a terminal, or when some option is given, etc.

- A client program can have any name. In these specification we assume it's called `tldr`.

- Basic usage: `tldr NAME...` - display the TLDR content for NAME, and exit.
  - TLDR content is markup text (SRC from now on) in a format specified below.
  - A procedure to resolve NAME to its SRC (e.g. the content of some file) is described below.
  - A client may transform SRC for display purpose in a way it finds useful, which may include both enhancements and reductions. Examples: stripping markup marks and empty lines, adding colors, indentations, creating an HTML and displaying it in a browser, etc.
  - If stdout is not a terminal, It's recommended to produce output which can be processed as text. A common practice in such case is that terminal escape sequences are stripped, formatting might be simpler, lists are produced with one item per line, etc.

And now, after it's clear WHAT it does, you start to specify HOW it does it.

- SRC:
  - SRC is the content of an object, identified by the string ID, which can be broken into DIR/FILE.
  - Retrieving SRC according to the string ID is up to the client. E.g. it could extract ID from a zip file, or read a file by that name under a local directory, or query a database table for an ID item, etc.

- Resolving ID happens in 3 stages:
  - Resolving one string FILE.
  - Resolving an ordered list of DIR strings.
  - For each DIR in the ordered list, try to retrieve the SRC of ID DIR/FILE.
    - On the first successful retrieval: display SRC and exit with success code.
    - If no retrieval was successful: display an error messge and exit with failure code.

- Resolving FILE:
  - The string FILE is created by concatenating all NAME items with `-` as separator, and appending `.md`.
    - E.g. `tldr foo` results in FILE: `foo.md`.
    - E.g. `tldr foo bar baz` results in FILE: `foo-bar-baz.md`.

- Creating the DIRs list:
  - The basic list of ordered DIR items is this exact list: `common`, `linux`, `osx`, `windows`, `sunos`.

A client is allowed to use only the basic list, and continue to resolve SRC according FILE and this list alone.

Alternatively, a client may try to improve the suitability of the list to the current environment by identifying the current platform and/or preferred languages, as follows:

Platform:
- If - and ONLY if - the client can clearly identify the current environment as one or more of `linux`, `osx`, `windows`, `sunos`:
  - It may move identified items to the head of the list - to be attempted before `common`.
  - (recommended to NOT remove platforms after common).
  - (...).

Languages:
- (how to create a list of language strings).
- (how use the languages list together with the platforms list for a combined DIRs list).

Then, once it's clear what is the main function and behavior, you specify options - which are optional.

**Comment by sbrl** [comment]
Wow, that's a lot of text here. It will take me a while to read through it all.

**Comment by sbrl** [comment]
> POSIX has a great set of recomendations

Oh nice! Perhaps we could adapt some of that then to enhance the clarity of our spec.

---------------

> But if only the structure changes while the names remain in the same format as now, then for instance the locale it_IT would not find any pages, because the only Italian pages are in pages.it and not pages.it_IT.

Hrm, that's an awkward one. I wonder what we do currently with that? Anyone have any ideas (/cc @owenvoke or @agnivade perhaps)?

---------------

> Not right now. 

Fair enough. You've raised a lot of individual points here though - so I'm unsure as to whether I'll be able to remember and correct them all. You'll have to remind us when the next PR goes through :P

---------------

> I'm still not entirely clear on the language-to-directory resolution clients are expected to use. Do you prefer that I open a new issue for it?

Yes please :-)

---------------

> and they go "wait, WHAT?!?", was I supposed to run the command NAME?

Ah, interesting interpretation there. I've started to revise the spec to aid clarity a bit there - so hopefully that clears that up.

--------------------

> And another subject, it's technically impossible to "detect" the current platform without knowing IN ADVANCE the exact list of platforms which you have pages for - and even that requires considerable guessing of matching the current platform to one or more of the platforms which you support.

Hrm. I see what you're saying there, but note this part of the spec:

> Clients MAY NOT support new platforms (though such support is RECOMMENDED), but MUST NOT break if additional platforms are added

---------------

Ok, so I've pushed some improvements to the spec to a new branch, but we need to wait for #4246 to be merged before we can open a PR and start working some things out.

Particularly with respect to some of the changes here, I think it might be best if you open a separate PR after my changes have gone through. That way you can demonstrate what you mean - which would be more effective that us going back and forth here or in a PR where I've attempted to take action.

**Comment by avih** [comment]
> Wow, that's a lot of text here. It will take me a while to read through it all.

Yeah, sorry about that. Take your time ;)

> > I'm still not entirely clear on the language-to-directory resolution clients are expected to use. Do you prefer that I open a new issue for it?
> 
> Yes please :-)

> > But if only the structure changes while the names remain in the same format as now, then for instance the locale it_IT would not find any pages, because the only Italian pages are in pages.it and not pages.it_IT.
> 
> Hrm, that's an awkward one. I wonder what we do currently with that? Anyone have any ideas (/cc @owenvoke or @agnivade perhaps)?

This "awkwardness" was my main question. In other words, the algorithm doesn't work for the languages which you currently have pages for, so you clearly understand the issue which I was wondering about. Feel free to resolve it anywhere you prefer.


> Ah, interesting interpretation there. I've started to revise the spec to aid clarity a bit there - so hopefully that clears that up.

Thanks. I'll try to check it out soon.

> > And another subject, it's technically impossible to "detect" the current platform without knowing IN ADVANCE...
> 
> Hrm. I see what you're saying there, but note this part of the spec: ...

This only partially covers it:
- You give `windows`, `osx` etc as "examples", but examples are not enough here - you should describe the exact and full list of platforms which a client should support (with linux and sunos).
- This still leaves the awkward guesswork a client MUST do (your requirement) in order to match the current platform to one of the listed/specified platforms - even if they're known in advance. Anything which involves guesswork just cannot be a "MUST" - it's an impossible requirement, unless you describe exactly how is a client supposed to make this guess, and what it should do if it can't guess.



> Ok, so I've pushed some improvements to the spec to a new branch, but we need to wait for #4246 to be merged before we can open a PR and start working some things out.

Yeah, the exit code is kinda obvious. I wasn't aware of this issue but I also specified the exit code behavior at the "spec" I posted above.


> Particularly with respect to some of the changes here, I think it might be best if you open a separate PR after my changes have gone through. That way you can demonstrate what you mean - which would be more effective that us going back and forth here or in a PR where I've attempted to take action.

One step at a time. I'm not following the spec daily, so if you want me to help review some changes, just ping me.

**Comment by agnivade** [comment]
> Hrm, that's an awkward one. I wonder what we do currently with that? Anyone have any ideas (/cc @owenvoke or @agnivade perhaps)?

Perhaps we can do a fallback to check for `it` if there's no `it_IT`.

**Comment by avih** [comment]
> Perhaps we can do a fallback to check for it if there's no it_IT.

That's what other clients (including node.js) do, but to quote @sbrl:
> Hrm, that approach has a number of issues. Regional dialects can have significant variances, and the resolution order for different languages is already clearly defined in the languages section of the spec

which I don't disagree with, but when confronted with the reality of existing directories which you have, this part of the spec becomes mostly useless.

**Comment by Managor** [comment]
@kbdharun You're more informed on the client specifications. Are there any points in this thread that are missing in the current specs?

**Comment by kbdharun** [comment]
> [@kbdharun](https://github.com/kbdharun) You're more informed on the client specifications. Are there any points in this thread that are missing in the current specs?

Thanks for notifying me again about this issue today, most of these suggestions has been implemented in subsequent specification releases like v1.4, v1.5, v2.0, v2.1, v2.2, v2.3 (latest). 

The only thing I see missing/not mention in spec clearly is fallback to a base local in case a dialect isn't present i.e. `ta` (Tamil which also known as `ta_IN`  - Tamil [India]) page should be shown when a user's local language is set to `ta_LK` (Tamil [Sri Lanka]) since we don't have localized pages for the dialect. Let's open a discussion for this in a newer issue and close this one since a lot of information has become outdated. Once again thanks to everyone who participated in the issue to help improve the client spec.
